config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java

```diff
@@ -74,22 +73,66 @@
-                                final BundleMetadata bundleMetadata = bundleMetadataBuilder.build(annotatedBundle, annotatedBundle.getAnnotatedEntity(),
-                                        entities, projectGroupName, projectVersion);
-                                annotatedElements.put(annotatedBundle.getBundleName(), ImmutablePair.of(annotatedElement,
-                                        bundleMetadata));
+
+                                // Create deployment bundle
+                                final Element bundleElement = bundleDocumentBuilder.build(document, entities);
+
+                                // Create DELETE bundle - ALWAYS skip environment entities
+                                final Element deleteBundleElement = createDeleteBundle(document, entities, bundle,
+                                        annotatedEntity);
+
+                                // Create bundle metadata
+                                final BundleMetadata bundleMetadata = bundleMetadataBuilder.build(annotatedBundle,
+                                        annotatedEntity, entities, projectGroupName, projectVersion);
+
+                                annotatedElements.put(annotatedBundle.getBundleName(),
+                                        new BundleArtifacts(bundleElement, deleteBundleElement, bundleMetadata));
                             }
                         })
         );
 
         return annotatedElements;
     }
 
-    private void loadEntityDependencies(String policyNameWithPath, AnnotatedBundle annotatedBundle, Bundle bundle) {
+    private Element createDeleteBundle(final Document document, List<Entity> entities, final Bundle bundle,
+                                       final AnnotatedEntity<GatewayEntity> annotatedEntity) {
+        List<Entity> deleteBundleEntities = new ArrayList<>(entities);
+
+        // If @redeployable annotation is added, we can blindly include all the dependencies in the DELETE bundle.
+        // Else, we have to include only non-reusable entities
+        if (!annotatedEntity.isRedeployable()) {
+            // Include only non-reusable entities
+            AnnotatedBundle annotatedBundle = new AnnotatedBundle(bundle, annotatedEntity);
+            Map bundleEntities = annotatedBundle.getEntities(annotatedEntity.getEntity().getClass());
+            bundleEntities.put(annotatedEntity.getEntityName(), annotatedEntity.getEntity());
+            loadEntityDependencies(annotatedEntity.getPolicyName(), annotatedBundle, bundle, true);
+
+            Iterator<Entity> it = deleteBundleEntities.iterator();
+            while (it.hasNext()) {
+                final Entity entity = it.next();
+                final Class<? extends GatewayEntity> entityClass = entityTypeRegistry.getEntityClass(entity.getType());
+                Map entityMap = annotatedBundle.getEntities(entityClass);
+                if (!entityMap.containsKey(entity.getName())) {
```

Comment:
This might fail for policy as key in the map contains policy path

Reply:
just verify this for subfolder policy

Reply:
It's working. Key for entity map is actually only the policy name (not with path)